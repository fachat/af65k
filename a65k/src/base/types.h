/****************************************************************************

    type definitions
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef TYPES_H
#define TYPES_H

#include <stdbool.h>

typedef struct _type type_t;

// more or less a class definition type
struct _type {
	const char 	*name;
	int		sizeoftype;
};

typedef bool 	bool_t;

// because the standard isprint() does not handle UTF8 codes
static inline bool_t isPrint(char c) {
	return c > 31;
}

#endif

