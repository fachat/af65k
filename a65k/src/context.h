
/****************************************************************************

    context management
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

// The context is what describes the current assembler context.
// It contains information about the source file, the current segment,
// the current CPU and its CPU modes (e.g. 65816 index register width,
// or 65k width)

#ifndef CONTEXT_H
#define CONTEXT_H


typedef struct context_s context_t;

struct context_s {
	const segment_t	*segment;
	const cpu_t	*cpu;
	int		cpu_width;	// CPU width, normally taken from *cpu, but can be modified with .width
        bool_t          index_width;    // true when index registers are wide in 65816 (.xe/.xs)
        bool_t          acc_width;      // true when accumulator is wide in 65816 
};

// create the initial context. Usually only called at beginning of parse
const context_t *context_init(const segment_t *segment, const cpu_t *cpu);

// duplicates context, so attributes can be modified. The new context becomes the 
// current context.
context_t *context_dup();

// return the current context
const context_t *context();

#endif

