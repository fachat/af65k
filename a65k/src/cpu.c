
/****************************************************************************

    CPU management
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <string.h>

#include "mem.h"
#include "hashmap.h"
#include "astring.h"
#include "cpu.h"

#include "infiles.h"
#include "errors.h"


static const cpu_t cpus[] = {
	{ CPU_NMOS,
		"nmos",			ISA_NMOS_ALL,	
		W_WORD,	true, 	false,
		"Standard NMOS 6502, with Decimal mode and illegal opcodes." },
	{ CPU_NMOS_NO_ILLEGAL,
		"nmos_no_illegal",	ISA_NMOS + ISA_NMOS_BCD,	
		W_WORD,	true, 	true,
		"Standard NMOS 6502 with decimal mode but without illegal opcodes." },
	{ CPU_CMOS,
		"cmos",			ISA_NMOS + ISA_CMOS,	
		W_WORD,	true, 	false,
		"Standard 65C02 CMOS, without the Rockwell extensions." },
	{ CPU_RCMOS,
		"cmos_rockwell",	ISA_NMOS + ISA_NMOS_BCD + ISA_CMOS,	
		W_WORD,	true, 	false,
		"R65C02 Rockwell CMOS with the RMB/SMB/BBR/BBS extensions." },
	{ CPU_NO_BCD,
		"nmos_no_bcd",		ISA_NMOS + ISA_NMOS_ILLEGAL,	
		W_WORD,	false, 	false,
		"NMOS 6502 with illegal opcodes, but no decimal mode (SED)." },
	{ CPU_NO_BCD_NO_ILLEGAL,
		"nmos_no_illegal_no_bcd", ISA_NMOS,	
		W_WORD,	false,	true,
		"NMOS 6502 without illegal opcodes, and no decimal mode." },
	{ CPU_802,
		"65802",		ISA_816_ALL,
		W_816,	true,	false,
		"WDC65802" },
	{ CPU_816,
		"65816",		ISA_65K_ALL,	
		W_816,	true, 	false,
		"WDC65816" },
	{ CPU_65K,
		"65k",			ISA_65K_ALL,	
		W_QUAD,	true, 	false,
		"Generic 65K CPU" },
	{ CPU_65K_W,
		"65k_w",		ISA_65K_ALL,	
		W_WORD,	true, 	false,
		"65K CPU in 16 bit width." },
	{ CPU_65K_L,
		"65k_l",		ISA_65K_ALL,	
		W_LONG,	true, 	false,
		"65K CPU in 32 bit width." },
	{ CPU_65K_Q,
		"65k_q",		ISA_65K_ALL,	
		W_QUAD,	true, 	false,
		"65K CPU in 64 bit width." },
};

static hash_t *cpu_typemap = NULL;

static const void *cpu_key_from_entry(const void *entry) {
	cpu_t *cpu = (cpu_t*) entry;
	return cpu->name;
}

void cpu_module_init() {

	int num_cpus = sizeof(cpus)/sizeof(cpu_t);

	cpu_typemap = hash_init_stringkey(10, 5, cpu_key_from_entry);

	for (int i = 0; i < num_cpus; i++) {
		hash_put(cpu_typemap, (void*)&cpus[i]);
	}
}

const cpu_t *cpu_by_name(const char *name) {

	cpu_t *cpu = (cpu_t*) hash_get(cpu_typemap, name);

	return cpu;
}

const cpu_t *cpu_by_type(cpu_type type) {

	unsigned int num_cpus = sizeof(cpus)/sizeof(cpu_t);
	if (type >= num_cpus) {
		error_illegal_cpu_type(type, num_cpus);
		return NULL;
	}
	return &cpus[type];
}



