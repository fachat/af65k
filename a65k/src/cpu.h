
/****************************************************************************

    CPU management
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef CPU_H
#define CPU_H

typedef enum {
        W_NONE                  = 0,
        W_8                     = 1,
        W_16                    = 2,
        W_24                    = 4,
        W_32                    = 8,
        W_64                    = 16,
} width_map;

// possible register widths (i.e. W_24 3-byte addresses are missing)
#define W_ALLREG        (W_8 + W_16 + W_32 + W_64)
#define W_816           (W_8 + W_16 + W_24)
#define W_BYTE          (W_8)
#define W_WORD          (W_8 + W_16)
#define W_LONG          (W_8 + W_16 + W_32)
#define W_QUAD          (W_8 + W_16 + W_32 + W_64)



typedef enum {
        ISA_NMOS                = 1,    // standard NMOS without SED
        ISA_NMOS_ILLEGAL        = 2,    // NMOS illegal opcodes
        ISA_NMOS_BCD            = 4,    // NMOS SED opcode
        ISA_CMOS                = 8,    // CMOS extensions
        ISA_CMOS_ROCKWELL       = 16,   // RMB/SMB/BBS/BBR opcodes
        ISA_816                 = 32,   // 65816 extensions to CMOS 6502
        ISA_65K                 = 64,   // 65K extensions to CMOS 6502
} isa_map;

#define ISA_NMOS_ALL    (ISA_NMOS + ISA_NMOS_ILLEGAL + ISA_NMOS_BCD)
#define ISA_CMOS_ALL    (ISA_NMOS + ISA_CMOS + ISA_CMOS_ROCKWELL)
#define ISA_816_ALL     (ISA_NMOS + ISA_CMOS + ISA_816)
#define ISA_65K_ALL     (ISA_NMOS + ISA_CMOS + ISA_65K)
#define ISA_ALL         (ISA_NMOS + ISA_NMOS_ILLEGAL + ISA_NMOS_BCD + ISA_CMOS + ISA_CMOS_ROCKWELL + ISA_816 + ISA_65K)


typedef enum {
	CPU_NMOS		= 0,	// legal NMOS opcodes, decimal mode and illegal opcodes
	CPU_NMOS_NO_ILLEGAL	= 1,	// legal NMOS opcodes incl. decimal mode, but no illegal opcodes
	CPU_CMOS		= 2,	// CMOS opcodes (without Rockwell extensions)
	CPU_RCMOS		= 3,	// CMOS plus Rockwell extensions
	CPU_NO_BCD		= 4,	// NMOS with illegal opcodes, but without BCD mode (e.g. NES)
	CPU_NO_BCD_NO_ILLEGAL	= 5,	// NMOS without illegal opcodes, and without BCD mode
	CPU_802			= 6,	// 65802
	CPU_816			= 7,	// 65816
	CPU_65K			= 8,	// base 65k CPU (implicitely 64 bit, modifyable with .width) 
	CPU_65K_W		= 9,	// 65k CPU 16-bit width
	CPU_65K_L		= 10,	// 65k CPU 32-bit width
	CPU_65K_Q		= 11,	// 65k CPU 64-bit width
} cpu_type;

typedef struct {
	const cpu_type	type;
	const char 	*name;
	const isa_map	isa;
	const width_map	width;		// memory model size 16, 24, 32, 64
	const bool_t	has_bcd;
	const bool_t	has_illegal;
	const char 	*desc;		// description of CPU model
} cpu_t;


void cpu_module_init();

const cpu_t *cpu_by_name(const char *name);

const cpu_t *cpu_by_type(cpu_type type);


#endif

