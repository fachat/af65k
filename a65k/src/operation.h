
/****************************************************************************

    CPU management
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef OPERATION_H
#define OPERATION_H



typedef enum {
	SY_IMP		= 0,	// implied
	SY_IMM		= 1,	// immediate
	
	SY_ABS		= 2,	// absolute/relative address: zp; word; long; quad; rel; relwide; rellong; relquad
	SY_ABSX		= 3,	// absolute x-indexed: zp,x; word,x; long,x; quad,x
	SY_ABSY		= 4,	// absolute y-indexed: zp,y; word,y; long,y; quad,y

	SY_IND		= 5,	// zp/word-indirect, word address: (zp); (word)
	SY_INDY		= 6,	// zp/word-indirect, y-indexed, word address: (zp),y; (word),y
	SY_XIND		= 7,	// x-indexed, zp/word-indirect, word address: (zp,x); (word,x)

	SY_INDL		= 8,	// zp/word-indirect, long address: [zp]; [word]
	SY_INDYL	= 9,	// zp/word-indirect, y-indexed, long address: [zp],y; [word],y
	SY_XINDL	= 10,	// x-indexed, zp/word-indirect, long address: [zp,x]; [word,x]

	SY_INDQ		= 11,	// zp/word-indirect, quad address: [[zp]]; [[word]]
	SY_INDYQ	= 12,	// zp/word-indirect, y-indexed, quad address: [[zp]],y; [[word]],y
	SY_XINDQ	= 13,	// x-indexed, zp/word-indirect, quad address: [[zp,x]]; [[word,x]]
	
	SY_MV		= 14,	// two byte 65816 MVN/MVP addressing
	SY_BBREL	= 15,	// zeropage address, plus relative address, R65C02 BBR/BBS
	
	SY_MAX		= 16,	// end

	// virtual addressing modes. Will be mapped to SY_IND, SY_INDL or SY_INDQ depending
	// on memory model used during assembly
	SY_INDD		= 17,	// zp/word-indirect, dynamic address: ((zp)); ((word))
	SY_INDYD	= 18,	// zp/word-zp-indirect, y-indexed, dynamic address: ((zp)),y; ((word)),y
	SY_XINDD	= 19,	// x-indexed, zp/word-indirect, dynamic address: ((zp,x)); ((word,x))

} syntax_type;


typedef enum {
	AM_NONE		= -1,

	AM_IMP		= 0,	// implied
	AM_IMM8		= 1,	// immediate
	AM_IMM16	= 2,	// immediate
	AM_IMM32	= 3,	// immediate
	AM_IMM64	= 4,	// immediate
	
	AM_ABS8		= 5,	// absolute/relative address: zp
	AM_ABS16	= 6,	// absolute/relative address: word
	AM_ABS24	= 7,	// absolute address: word+bank byte
	AM_ABS32	= 8,	// absolute/relative address: long
	AM_ABS64	= 9,	// absolute/relative address: quad

	AM_ABS8X	= 10,	// absolute x-indexed: zp,x
	AM_ABS16X	= 11,	// absolute x-indexed: word,x
	AM_ABS32X	= 12,	// absolute x-indexed: long,x
	AM_ABS64X	= 13,	// absolute x-indexed: quad,x

	AM_ABS8Y	= 14,	// absolute y-indexed: zp,y
	AM_ABS16Y	= 15,	// absolute y-indexed: word,y
	AM_ABS32Y	= 16,	// absolute y-indexed: long,y
	AM_ABS64Y	= 17,	// absolute y-indexed: quad,y

	AM_IND8		= 18,	// zp/word-indirect, word address: (zp)
	AM_IND16	= 19,	// zp/word-indirect, word address: (word)
	AM_IND8Y	= 20,	// zp/word-indirect, y-indexed, word address: (zp),y
	AM_IND16Y	= 21,	// zp/word-indirect, y-indexed, word address: (word),y
	AM_XIND8	= 22,	// x-indexed, zp/word-indirect, word address: (zp,x)
	AM_XIND16	= 23,	// x-indexed, zp/word-indirect, word address: (word,x)

	AM_IND8L	= 24,	// zp/word-indirect, long address: [zp]
	AM_IND16L	= 25,	// zp/word-indirect, long address: [word]
	AM_IND8YL	= 26,	// zp/word-indirect, y-indexed, long address: [zp],y
	AM_IND16YL	= 27,	// zp/word-indirect, y-indexed, long address: [word],y
	AM_XIND8L	= 28,	// x-indexed, zp/word-indirect, long address: [zp,x]
	AM_XIND16L	= 29,	// x-indexed, zp/word-indirect, long address: [word,x]

	AM_IND8Q	= 30,	// zp/word-indirect, quad address: [[zp]]
	AM_IND16Q	= 31,	// zp/word-indirect, quad address: [[word]]
	AM_IND8YQ	= 32,	// zp/word-indirect, y-indexed, quad address: [[zp]],y
	AM_IND16YQ	= 33,	// zp/word-indirect, y-indexed, quad address: [[word]],y
	AM_XIND8Q	= 34,	// x-indexed, zp/word-indirect, quad address: [[zp,x]]
	AM_XIND16Q	= 35,	// x-indexed, zp/word-indirect, quad address: [[word,x]]
	
	AM_MV		= 36,	// two byte 65816 MVN/MVP addressing
	AM_BBREL	= 37,	// zeropage address, plus relative address, R65C02 BBR/BBS
	
	AM_MAX		= 38,	// end
} amode_type;
	
typedef enum {
	PG_BASE		= 0,	// base page, no page prefix needed
	PG_EXT		= 1,	// extended page
	PG_QUICK	= 2,	// quick page
	PG_SYS		= 3,	// system page
} pg_type;


typedef struct {
	bool_t		is_valid;
	unsigned char	code;
	pg_type 	page_prefix;
} opcode_t;

// operation - equivalent to the mnemonic, like "lda", "adc", "inx", ...
typedef struct operation_s operation_t;
struct operation_s {
	const char	*name;
	const isa_map	isa;		// what ISA is it from? Also selects BCD and Illegal opcodes
	const bool_t	abs_is_rel;	// is branch with relative addressing?
	operation_t	*next;
	opcode_t	opcodes[AM_MAX];
};

typedef struct {
	bool_t		set_am_prefix;
	bool_t		rs_is_width;		// use RS-prefix for width
	pg_type		page_prefix;
	unsigned char	code;
} codepoint_t;

void operation_module_init();

const operation_t *operation_find(const char *name);

bool_t opcode_find(const position_t *loc, const cpu_t *cpu, const operation_t *op, syntax_type syntax, int opsize_in_bytes,
	codepoint_t *returned_code);

#endif

