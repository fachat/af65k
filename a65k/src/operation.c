
/****************************************************************************

    CPU operation management
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <string.h>

#include "mem.h"
#include "hashmap.h"
#include "astring.h"
#include "cpu.h"
#include "operation.h"


static struct { 
	cpu_type cpu; 
	isa_type isa;
	width_map width;
} cpu_isa[] = {
	{ CPU_NMOS,		ISA_NMOS + ISA_NMOS_BCD,			W_8 + W_16	},
	{ CPU_NMOS_ILLEGAL,	ISA_NMOS + ISA_NMOS_ILLEGAL + ISA_NMOS_BCD,	W_8 + W_16 	},
	{ CPU_CMOS,		ISA_NMOS + ISA_CMOS,				W_8 + W_16	},
	{ CPU_RCMOS,		ISA_NMOS + ISA_CMOS + ISA_CMOS_ROCKWELL,	W_8 + W_16	},
	{ CPU_NOBCD,		ISA_NMOS, 					W_8 + W_16	},
	{ CPU_NOBCD_ILLEGAL,	ISA_NMOS + ISA_NMOS_ILLEGAL,			W_8 + W_16	},
	{ CPU_802,		ISA_NMOS + ISA_CMOS + ISA_816,			W_8 + W_16 + W_24 },
	{ CPU_816,		ISA_NMOS + ISA_CMOS + ISA_816,			W_8 + W_16 + W_24 },
	{ CPU_65K,		ISA_NMOS + ISA_CMOS + ISA_65K,			W_8 + W_16 + W_32 + W_64 },
	{ CPU_65K_W,		ISA_NMOS + ISA_CMOS + ISA_65K,			W_8 + W_16 	},
	{ CPU_65K_L,		ISA_NMOS + ISA_CMOS + ISA_65K,			W_8 + W_16 + W_32 },
	{ CPU_65K_Q,		ISA_NMOS + ISA_CMOS + ISA_65K,			W_8 + W_16 + W_32 + W_64 },
	
};

void operation_module_init() {
}


operation_t *operation_find(const cpu_t *cpu, const char *name) {
	(void) cpu;
	(void) name;
	return NULL;
}

