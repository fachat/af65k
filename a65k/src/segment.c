
/****************************************************************************

    segment management
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include "mem.h"
#include "types.h"
#include "hashmap.h"
#include "position.h"
#include "cpu.h"
#include "segment.h"


// default names for the xa65 default segments
const char *SEG_ANY_NAME = "*";
const char *SEG_BSS_NAME = "bss";
const char *SEG_TEXT_NAME = "text";
const char *SEG_DATA_NAME = "data";
const char *SEG_ZP_NAME = "zp";

static hash_t *segments = NULL;

static type_t segment_memtype = {
	"segment_t",
	sizeof(segment_t)
};

static const void *segment_key_from_entry(const void *entry) {
	return ((const segment_t*)entry)->name;
}

void segment_module_init() {

	segments = hash_init_stringkey(5, 5, &segment_key_from_entry);
}

// create a new segment or find an existing, matching one
segment_t *segment_new(const position_t *loc, const char *name, seg_type type, cpu_type cpu, bool_t readonly) {

	segment_t *existing = hash_get(segments, name);

	if (existing == NULL) {
		existing = mem_alloc(&segment_memtype);

		existing->name = mem_alloc_str(name);
		existing->type = type;
		existing->readonly = readonly;
		existing->cpu_width = cpu_by_type(loc, cpu)->width;
	}

	return existing;
}


