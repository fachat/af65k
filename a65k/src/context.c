
/****************************************************************************

    context management
    Copyright (C) 2012,2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <stdio.h>
#include <string.h>

#include "types.h"
#include "infiles.h"
#include "array_list.h"
#include "hashmap.h"
#include "mem.h"
#include "cpu.h"
#include "segment.h"
#include "context.h"


#define	DEFAULT_SIZE_CHILDREN_BUCKET	25

static type_t context_memtype = {
	"context",
	sizeof(context_t)
};

static context_t *current_context = NULL;


// create a new context. Usually only called at beginning of parse
const context_t *context_init(const segment_t *segment, const cpu_t *cpu) {

	context_t *ctx = mem_alloc(&context_memtype);

	ctx->segment = segment;
	ctx->cpu = cpu;

	ctx->cpu_width = cpu->width;
	ctx->index_width = false;
	ctx->acc_width = false;

	current_context = ctx;

	return current_context;
}

// duplicates context, so attributes can be modified
context_t *context_dup() {


	context_t *newctx = mem_alloc(&context_memtype);

	context_t *oldctx = current_context;

	memcpy(newctx, oldctx, sizeof(context_t));

	current_context = newctx;

	return current_context;
}


const context_t *context() {
	return current_context;
}

