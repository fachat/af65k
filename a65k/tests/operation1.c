

#include <stdio.h>

#include "types.h"
#include "position.h"
#include "cpu.h"
#include "segment.h"
#include "context.h"
#include "operation.h"


int main(int argc, char *argv[]) {

	cpu_module_init();
	segment_module_init();
	operation_module_init();

	const cpu_t *cpu = cpu_by_name("nmos");

	const segment_t *seg = segment_new(NULL, "test", SEG_ANY, cpu->type, true);

	const context_t *ctx = context_init(seg, cpu);

	const operation_t *op = operation_find("adc");

	if (op == NULL) {
		printf("Opcode 'adc' not found!\n");
	}

	codepoint_t cp;

	bool_t rc = opcode_find(NULL, ctx, op, SY_IMM, 1, &cp);

	if (rc) {
		printf("Opcode 'adc' with addressing mode am=%d, rs_is_width=%d, prefix=%d, found: %02x\n",
			cp.set_am_prefix, cp.rs_is_width, cp.page_prefix, cp.code);
	}
}



