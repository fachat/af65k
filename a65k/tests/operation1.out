Opcode 'adc' with addressing mode am=0, rs_is_width=0, prefix=0, found: 63
